src/main/cpp/src/hyper_log_log_plus_plus.cu

```diff
@@ -0,0 +1,964 @@
+ *    (1,            1),          g1
+ * // outputs result at segemnt end for this thread to registers_thread_cache
+ *    (1,            9),          g1
+ * ---------- segment 1 begin --------------------------
+ *    (1,            1),          g1
+ *    (1,            1),          g1
+ *    (1,            5),          g1
+ * // outputs result at segemnt end for this thread to registers_thread_cache
+ *    (1,            1),          g1
+ * ---------- segment 2 begin --------------------------
+ *    (1,            1),          g1
+ *    (1,            1),          g1
+ *    (1,            8),          g1
+ * // outputs result at segemnt end for this thread to registers_thread_cache
+ * // assumes meets new group when at the end, save to registers_output_cache
+ *    (1,            1),          g1
+ * ]
+ * Output e.g.:
+ *
+ * group_labels_thread_cache:
+ * [
+ *   g1
+ *   g1
+ *   g1
+ * ]
+ * Has num_threads rows.
+ *
+ * registers_thread_cache:
+ * [
+ *    512 values: [0, 9, 0, ... ] // register values for group 1
+ *    512 values: [0, 5, 0, ... ] // register values for group 1
+ *    512 values: [0, 8, 0, ... ] // register values for group 1
+ * ]
+ * Has num_threads rows, each row is corresponding to
+ * `group_labels_thread_cache`
+ *
+ * registers_output_cache:
+ * [
+ *    512 values: [2, 0, 0, ... ] // register values for group 0
+ *    512 values: [0, 8, 0, ... ] // register values for group 1
+ * ]
+ * Has num_groups rows.
+ *
+ * The next kernel will merge the registers_output_cache and
+ * registers_thread_cache and get the final result.
+ */
+template <int num_hashs_per_thread>
+CUDF_KERNEL void partial_group_sketches_from_hashs_kernel(
+  cudf::column_device_view hashs,
+  cudf::device_span<cudf::size_type const> group_labels,
+  int64_t const precision,                          // num of bits for register addressing, e.g.: 9
+  int* const registers_output_cache,                // num is num_groups * num_registers_per_sketch
+  int* const registers_thread_cache,                // num is num_threads * num_registers_per_sketch
+  cudf::size_type* const group_labels_thread_cache  // save the group labels for each thread
+)
+{
+  auto const tid          = cudf::detail::grid_1d::global_thread_id();
+  int64_t const num_hashs = hashs.size();
+  if (tid * num_hashs_per_thread >= hashs.size()) { return; }
+
+  // 2^precision = num_registers_per_sketch
+  int64_t num_registers_per_sketch = 1L << precision;
+  // e.g.: integer in binary: 1 0000 0000
+  uint64_t const w_padding = 1ULL << (precision - 1);
+  // e.g.: 64 - 9 = 55
+  int const idx_shift = 64 - precision;
+
+  auto const hash_first = tid * num_hashs_per_thread;
+  auto const hash_end   = cuda::std::min((tid + 1) * num_hashs_per_thread, num_hashs);
+
+  // init sketches for each thread
+  int* const sketch_ptr = registers_thread_cache + tid * num_registers_per_sketch;
+  for (auto i = 0; i < num_registers_per_sketch; i++) {
+    sketch_ptr[i] = 0;
+  }
```

Comment:
I see that here (and throughout this function) we read/write to `sketch_ptr` serially. How big is the typical value of `num_registers_per_sketch`? If it is big like 512 as mentioned in the docs, can we process each element of `sketch_ptr` in one separate thread, to improve performance?

Reply:
Using more threads will probably improve the performance. I'll try it.

Reply:
There are 2 kinds of serially copies:
```cpp
  // initialize
  for (auto i = 0; i < num_registers_per_sketch; i++) {
    sketch_ptr[i] = 0;
  }
```

```cpp
  // meets a new group
      for (auto i = 0; i < num_registers_per_sketch; i++) {
        registers_output_cache[prev_group * num_registers_per_sketch + i] = sketch_ptr[i];
        sketch_ptr[i]                                                     = 0;
      }
```
One hash will be mapped to one address of 512 possibilities.
The first one is for initialization, seems it's not worth to have 512 threads to do initialization.
The second one is for meeting a new group. Usually the group number is less, so there is no significant copying.

Reply:
This is shorter:
```suggestion
  memset(sketch_ptr, 0, num_registers_per_sketch * sizeof(int));
```